### PR TITLE
Support named parameters in SQL HazelcastFunction

### PR DESCRIPTION
**Pull request by gierlachg**
Added support for named parameters to SQL `HazelcastFunction`.

A prerequisite for https://github.com/hazelcast/hazelcast-jet/pull/2944

### CODE REVIEW

File: hazelcast-sql-core/src/main/java/com/hazelcast/sql/impl/calcite/validate/operand/OperandCheckerProgram.java [hunk 1]
```diff
@@ -32,9 +40,11 @@ public OperandCheckerProgram(OperandChecker... checkers) {
     public boolean check(HazelcastCallBinding callBinding, boolean throwOnFailure) {
         boolean res = true;
 
+        boolean hasAssignment = hasAssignment(callBinding.getCall());
         for (int i = 0; i < checkers.length; i++) {
-            boolean checkerRes = checkers[i].check(callBinding, false, i);
+            OperandChecker checker = hasAssignment ? findCheckerByParamName(callBinding, i) : checkers[i];
```
Comment (viliam-durina) [inline hazelcast-sql-core/src/main/java/com/hazelcast/sql/impl/calcite/validate/operand/OperandCheckerProgram.java:45]: Can't we always find the checker by param name? Isn't it the more general way?

File: hazelcast-sql-core/src/main/java/com/hazelcast/sql/impl/calcite/validate/types/HazelcastTypeCoercion.java [hunk 2]
```diff
@@ -94,6 +95,22 @@ private boolean coerceNode(
...
 
+    private static SqlNode cast(SqlNode node, SqlDataTypeSpec targetTypeSpec) {
+        if (node.getKind() == SqlKind.ARGUMENT_ASSIGNMENT) {
+            // transform name => 'value' into name => cast('value')
+            SqlBasicCall call = (SqlBasicCall) node;
+
+            SqlNode value = call.getOperandList().get(0);
+            SqlNode name = call.getOperandList().get(1);
+
+            SqlNode cast = HazelcastSqlOperatorTable.CAST.createCall(SqlParserPos.ZERO, value, targetTypeSpec);
```
Comment (viliam-durina) [inline hazelcast-sql-core/src/main/java/com/hazelcast/sql/impl/calcite/validate/types/HazelcastTypeCoercion.java:106]: We can do a recursive call:
```suggestion
            SqlNode cast = cast(value, targetTypeSpec);
```

**Review by viliam-durina — APPROVED** [review]
Looks good, with 2 minor comments.

File: hazelcast-sql-core/src/main/java/com/hazelcast/sql/impl/calcite/validate/operand/OperandCheckerProgram.java (hunk 1 above)
Comment (gierlachg) [reply hazelcast-sql-core/src/main/java/com/hazelcast/sql/impl/calcite/validate/operand/OperandCheckerProgram.java:45]: When one uses positional arguments then there are no names (it's not an argument assignment) we could match params by.

Comment (viliam-durina) [reply hazelcast-sql-core/src/main/java/com/hazelcast/sql/impl/calcite/validate/operand/OperandCheckerProgram.java:45]: But the what about mixed named and non-named operands? The `hasAssignment` looks at all the operands, if any is named, we extract all of them by name.

Would also be nice if we could add a test in imdg - both for all named and mixed arguments.

Comment (gierlachg) [reply hazelcast-sql-core/src/main/java/com/hazelcast/sql/impl/calcite/validate/operand/OperandCheckerProgram.java:45]: Calcite (we) does not support mixed operands.

Comment (gierlachg) [reply hazelcast-sql-core/src/main/java/com/hazelcast/sql/impl/calcite/validate/operand/OperandCheckerProgram.java:45]: Calcite (we) does not support mixed operands.

File: hazelcast-sql-core/src/main/java/com/hazelcast/sql/impl/calcite/validate/operand/AbstractOperandChecker.java [hunk 3]
```diff
@@ -35,7 +35,7 @@ protected AbstractOperandChecker() {
     public boolean check(HazelcastCallBinding callBinding, boolean throwOnFailure, int operandIndex) {
         HazelcastSqlValidator validator = callBinding.getValidator();
 
-        SqlNode operand = callBinding.operand(operandIndex);
+        SqlNode operand = callBinding.getCall().getOperandList().get(operandIndex);
```
Comment (devozerov) [inline hazelcast-sql-core/src/main/java/com/hazelcast/sql/impl/calcite/validate/operand/AbstractOperandChecker.java:38]: What is the motivation for this change? I presume there is something wrong with `operand(operandIndex)`?

Comment (gierlachg) [reply hazelcast-sql-core/src/main/java/com/hazelcast/sql/impl/calcite/validate/operand/AbstractOperandChecker.java:38]: `operand()` returns _permuted_ node - because `OperandChecker`s operate on indices and positional arguments it would mess things up in case of argument assignments are present.

File: hazelcast-sql-core/src/main/java/com/hazelcast/sql/impl/calcite/validate/operand/OperandCheckerProgram.java [hunk 4]
```diff
@@ -46,4 +56,31 @@ public boolean check(HazelcastCallBinding callBinding, boolean throwOnFailure) {
...
+    private boolean hasAssignment(SqlCall call) {
+        for (SqlNode operand : call.getOperandList()) {
+            if (operand != null && operand.getKind() == SqlKind.ARGUMENT_ASSIGNMENT) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private OperandChecker findCheckerByParamName(SqlCallBinding binding, int index) {
```
Comment (devozerov) [inline hazelcast-sql-core/src/main/java/com/hazelcast/sql/impl/calcite/validate/operand/OperandCheckerProgram.java:69]: I couldn't really understand what this method does. The `OperandCheckerProgram` could be viewed as a convenient wrapper that executes predefined checkers for every parameter one by one. With this change, we start doing something completely different when there is a parameter. Could you please clarify a couple of things?
1. What is the purpose of the argument assignment operator? Can it be used in any context, or only in some designated operators, like table functions? 
2. What does `HazelcastFunction.getParamNames()` return? There is not JavaDoc and no implementation in IMDG, so the semantics is not clear.

Comment (gierlachg) [reply hazelcast-sql-core/src/main/java/com/hazelcast/sql/impl/calcite/validate/operand/OperandCheckerProgram.java:69]: 1. An alternative way of calling a function. Instead of positional arguments, one can use named ones, like `name => 'value'`. In theory, every operator could support named parameters.
2. Well, parameter names. It is used by Calcite whenever argument assignment is used.

Comment (devozerov) [reply hazelcast-sql-core/src/main/java/com/hazelcast/sql/impl/calcite/validate/operand/OperandCheckerProgram.java:69]: Got it, thanks.

**Review by devozerov — APPROVED** [review]

**Comment by gierlachg** [comment]
Trimmed original scoped of the PR a bit. Realized that `OperandCheckerProgram` should not deal with named arguments.
Left coercion support for argument assignment & couple of fixes regarding _getting_ operands (`callBinding.operand()` -> `callBinding.getCall().operand()`).

**Comment by gierlachg** [comment]
run-lab-run
